DataMigration/src/main/java/org/sagebionetworks/dian/datamigration/tools/schedulev2/ScheduleV2Migration.java

```diff
@@ -0,0 +1,335 @@
+import org.sagebionetworks.dian.datamigration.HmDataModel;
+import org.sagebionetworks.dian.datamigration.PathsHelper;
+import org.sagebionetworks.dian.datamigration.tools.adherence.CompletedTestV2;
+import org.sagebionetworks.dian.datamigration.tools.adherence.SageEarningsControllerV2;
+import org.sagebionetworks.dian.datamigration.tools.adherence.SageScheduleController;
+import org.sagebionetworks.dian.datamigration.tools.adherence.SageUserClientData;
+import org.sagebionetworks.dian.datamigration.tools.adherence.SageV1Schedule;
+import org.sagebionetworks.dian.datamigration.tools.adherence.SageV1StudyBurst;
+import org.sagebionetworks.dian.datamigration.tools.adherence.SageV2Availability;
+import org.sagebionetworks.dian.datamigration.tools.adherence.ScheduledSessionStart;
+import org.sagebionetworks.dian.datamigration.tools.adherence.earnings.EarningDetails;
+import org.sagebionetworks.dian.datamigration.tools.rescheduler.TestSchedule;
+
+import java.io.IOException;
+import java.nio.file.Path;
+import java.nio.file.Paths;
+import java.util.ArrayList;
+import java.util.HashMap;
+import java.util.HashSet;
+import java.util.List;
+import java.util.Map;
+import java.util.SortedSet;
+import java.util.TreeSet;
+
+import static java.util.Comparator.*;
+
+public class ScheduleV2Migration {
+
+    public static ObjectMapper objectMapper = new ObjectMapper();
+    public static Gson gson = new Gson();
+    public static SageScheduleController controller = new SageScheduleController();
+
+    public static void main(String[] args) throws IOException {
+        BridgeJavaSdkUtil.initialize(args[0], args[1], args[2]);
+        runV2Migration();
+    }
+
+    public static void runV2Migration() throws IOException {
+
+        List<Study> studyList = BridgeJavaSdkUtil.getAllStudies();
+        for (Study study : studyList) {
+            String studyId = study.getIdentifier();
+            HashSet<String> arcIdList = BridgeJavaSdkUtil.getArcIdsInStudy(studyId);
+
+            // wbfxxk is the ScheduleTemplate study, it is a test study for figuring out the V2 schedule
+            // Once we are ready to deploy this for all studies, use a curated list of all Study IDs
+            for (String arcId : arcIdList) {
+                StudyParticipant p = BridgeJavaSdkUtil.getParticipantByExternalId(arcId);
+
+                if (p.getStudyIds() == null || p.getStudyIds().isEmpty()) {
+                    System.out.println(arcId + " has withdrawn");
+                    continue; // user has withdrawn, no need to migrate
+                }
+
+                String uId = p.getId();
+                String sId = p.getStudyIds().get(0);
+
+                // Check for a user that has already migrated to V2 and signed into the app
+                SageUserClientData clientData = SageUserClientData.Companion.fromStudyParticipant(gson, p);
+                if (SageUserClientData.Companion.hasMigrated(clientData)) {
```

Comment:
Where does hasMigratedToV2 on SageUserClientData get set to true when the migration is done?

Reply:
So for a user that is still using the old version of the app, the migrator may run multiple times until they eventually log-in with the new app. Could data be missed if the migrator runs, then a user does an activity with the old app, and then updates to the new app, before the migrator has a chance to run again?

Reply:
I feel like this would be rare, but is a possibility.  This code will made into a docker image that runs nightly.  If they are currently within a study burst, while this happens, and they update the app in the middle of their testing day, it's possible a test or two may not be transferred over to adherence. 

Do you think it would be safer to always update their adherence data, no matter if the user already signed into the new app?  I wouldn't want to keep updating their schedule, because that risks overwriting any new study burst dates changes they make.

Reply:
Assuming that the timestamps match exactly, it should be safe to update their adherence data over and over. You could add a flag so that the migrator only runs 1 more time after the client has been updated.

Reply:
I also think it should be safe to keep running the updateAdherenceRecords as the completedOn date the function uses for each record is static and won't change. 

Let me make this change and test it out, to see if there are any issues.

I don't really think it is worth it to try and manage the "update one more time" flag if we assume that the updateAdherenceRecords function is safe.

Reply:
I updated the code to continue writing adherence records even after a user has migrated.  I migrated a test user that had study burst 1 and 2 data, then completed a test in their 4th study burst, and then re-ran the migration code a few times.

Re-running the migration did not cause any issues and the dashboard displayed the same information.  I also checked the adherence record list, and it didn't create duplicates for the records.

![Uploading Screen Shot 2022-09-26 at 18.23.05.png…]()

Reply:
That looks good.